File: rxbinding-leanback-v17/src/main/java/com/jakewharton/rxbinding2/support/v17/leanback/widget/SearchBarSearchQueryChangeEventsOnSubscribe.java
package com.jakewharton.rxbinding2.support.v17.leanback.widget;

import android.support.v17.leanback.widget.SearchBar;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class SearchBarSearchQueryChangeEventsOnSubscribe
    extends Observable<SearchBarSearchQueryEvent> {
  final SearchBar view;

  SearchBarSearchQueryChangeEventsOnSubscribe(SearchBar view) {
    this.view = view;
  }

  @Override
  protected void subscribeActual(final Observer<? super SearchBarSearchQueryEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setSearchBarListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements SearchBar.SearchBarListener {
    final SearchBar view;
    final Observer<? super SearchBarSearchQueryEvent> observer;

    Listener(SearchBar view, Observer<? super SearchBarSearchQueryEvent> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onSearchQueryChange(String query) {
      if (!isDisposed()) {
        observer.onNext(SearchBarSearchQueryChangedEvent.create(view, query));
      }
    }

    @Override public void onSearchQuerySubmit(String query) {
      if (!isDisposed()) {
        observer.onNext(SearchBarSearchQuerySubmittedEvent.create(view, query));
      }
    }

    @Override public void onKeyboardDismiss(String query) {
      if (!isDisposed()) {
        observer.onNext(SearchBarSearchQueryKeyboardDismissedEvent.create(view, query));
      }
    }

    @Override protected void onDispose() {
      view.setSearchBarListener(null);
    }
  }
}

File: rxbinding-leanback/src/main/java/com/jakewharton/rxbinding3/leanback/SearchBarSearchQueryChangeEventsOnSubscribe.kt
@file:JvmName("RxSearchBar")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.leanback

import androidx.annotation.CheckResult
import androidx.leanback.widget.SearchBar
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of [search query events][SearchBarSearchQueryEvent] on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun SearchBar.searchQueryChangeEvents(): Observable<SearchBarSearchQueryEvent> {
  return SearchBarSearchQueryChangeEventsOnSubscribe(this)
}

private class SearchBarSearchQueryChangeEventsOnSubscribe(
  private val view: SearchBar
) : Observable<SearchBarSearchQueryEvent>() {

  override fun subscribeActual(observer: Observer<in SearchBarSearchQueryEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.setSearchBarListener(listener)
  }

  private class Listener(
    private val view: SearchBar,
    private val observer: Observer<in SearchBarSearchQueryEvent>
  ) : MainThreadDisposable(), SearchBar.SearchBarListener {

    override fun onSearchQueryChange(query: String) {
      if (!isDisposed) {
        observer.onNext(SearchBarSearchQueryChangedEvent(view, query))
      }
    }

    override fun onSearchQuerySubmit(query: String) {
      if (!isDisposed) {
        observer.onNext(SearchBarSearchQuerySubmittedEvent(view, query))
      }
    }

    override fun onKeyboardDismiss(query: String) {
      if (!isDisposed) {
        observer.onNext(SearchBarSearchQueryKeyboardDismissedEvent(view, query))
      }
    }

    override fun onDispose() {
      view.setSearchBarListener(null)
    }
  }
}
